src/ess.jl

```diff
@@ -187,10 +178,11 @@ function mean_autocov(k::Int, cache::BDAESSCache)
     # compute mean autocovariance
     n = niter - k
     idxs = 1:n
+    idxs2 = k .+ idxs
     s = Statistics.mean(1:nchains) do j
-        return sum(idxs) do i
-            abs2(samples[i, j] - samples[k + i, j])
-        end
+        s1 = view(samples, idxs, j)
+        s2 = view(samples, idxs2, j)
+        return LinearAlgebra.dot(s1, s1) + LinearAlgebra.dot(s2, s2) - 2 * LinearAlgebra.dot(s1, s2)
```

Comment:
I assume that in contrast to the previous implementation this can return negative values (similar to what I experienced with Distances: https://github.com/JuliaStats/Distances.jl/pull/161). So we should change this to
```suggestion
        r = LinearAlgebra.dot(s1, s1) + LinearAlgebra.dot(s2, s2) - 2 * LinearAlgebra.dot(s1, s2)
        return max(0, r)
```

In general, I am surprised that this is actually faster than the current implementation. In particular the allocations seem weird, there should be no difference here (or at least the current implementation should not be worse). I think this needs some more benchmarks and more thorough investigation (probably it would also be good to only benchmark this function and not `ess_rhat`).

Reply:
I don't know where the allocations come from. But I realized just marking the inner loop of `sum` with `@inbounds` actually gave us more of a speed-up than using `dot`.